staging/src/k8s.io/client-go/tools/cache/delta_fifo.go

```diff
@@ -650,6 +656,35 @@ func (f *DeltaFIFO) syncKeyLocked(key string) error {
 	return nil
 }
 
+// A partial repair function.  Removes keys from f.queue that
+// correspond to nothing in f.items.  Removes keys from both when the
+// entry in f.items has zero length.  Ensures each key appears at most
+// once in queue.  Must be called with the lock held.
+func (f *DeltaFIFO) filterQueueLocked() {
```

Comment:
Just to share my two cents.

Personally, I think adding 29 lines of code (3.7% of the entire file) in a place which is supposed to be dead code is too much maintenance overhead. Any new reader of this class will have to try to understand what this function does. 

We often say we spent 80% of efforts in places which only occur 20% of the time, which is a big waste. I think this is one of such places. We have to draw a balance between 100% of code correctness and the increased maintenance overhead. 

I would rather not to pay the maintenance cost today for some bug which doesn't yet exist. Therefore, my vote for now is to simply return an error if dead code does get triggered due to a bug. We can fix the bug after with sufficient logging. 

Thoughts?

Reply:
I don't like a lot of dead code either.  So I revised again, this time returning an error when in a tight spot.  This does not avoid breaking the data structure, and introduces more possible error returns (so more previously-unexercised logic in the callers (all the clients of delta_fifo) may get exercised in that case).